wolfcrypt/src/cmac.c

```diff
@@ -305,6 +305,58 @@ int wc_CmacFinal(Cmac* cmac, byte* out, word32* outSz) {
     return ret;
 }
 
+
+int wc_AesCmacGenerate_ex(Cmac* cmac,
+                          byte* out, word32* outSz,
+                          const byte* in, word32 inSz,
+                          const byte* key, word32 keySz,
+                          int type, void* heap, int devId)
+{
+    int ret;
+
+    if ((out == NULL) || (in == NULL && inSz > 0))  {
+        return BAD_FUNC_ARG;
+    }
+
+#ifdef WOLF_CRYPTO_CB
+    /* Set devId regardless of value (invalid or not) */
+    cmac->devId = devId;
+    #ifndef WOLF_CRYPTO_CB_FIND
+    if (devId != INVALID_DEVID)
+    #endif
+    {
+        cmac->devCtx = NULL;
+
+        ret = wc_CryptoCb_Cmac(cmac, key, keySz, in, inSz, out, outSz,
+                type, NULL);
+        if (ret != CRYPTOCB_UNAVAILABLE)
+            return ret;
+        /* fall-through when unavailable */
+    }
+#endif
+
+    /* Clear CRYPTOCB_UNAVAILABLE return code */
+    ret = 0;
+
+    /* Init step is optional */
+    if (key != NULL) {
+        ret = wc_InitCmac_ex(cmac, key, keySz, type, NULL, heap, devId);
+    }
+    if (ret == 0) {
+        ret = wc_CmacUpdate(cmac, in, inSz);
+        /* Ensure we are freed and zeroed if not calling wc_CmacFinal */
+        if (ret != 0) {
```

Comment:
So the original generate function wasn't suitable for async use, since it allocated/freed on entry/exit, and even though this new generate_ex version takes a context, I still don't think its async-capable as it would still be calling "init" on every invocation. 

I added this line because you asked for the context to be zeroed in an earlier comment and I figured if we were completely clearing the context (e.g. zeroing the aes struct) then we probably wanted to free the AES substruct as well. But I wasn't thinking about this function being used in an async context, given that it is a "one shot". 

Also, the CMAC context doesn't hold any sort of async state, only the underlying AES module. So there is no way to "poll" the underlying state of the AES operation and resume using this function.

Given this, do you still want me to clear it on failure but not free? Or if you think it should be not async capable, then is the free okay?

Reply:
I think we need to thin about if/how cryptocb could support async responses.  Keep the code the same for now.